man/man1/adddrain.1
.TH adddrain 1 "USER COMMANDS"
.SH NAME
adddrain \- add river basin to coordinate file
.SH SYNOPSIS
.B adddrain
[\-longheader] [\-downstream] [\-short]
.I coord_all.clm coord.clm drainage.clm coord_basin.clm
.SH DESCRIPTION
Program creates new coordinate file with cells added that are part of the river basin. This has to be done for simulations with river routing enabled.
.SH OPTIONS
.TP
\-longheader
Long (version 2) header assumed for CLM files.
.TP
\-downstream
By default only cell from upstream are added. Enabling this flags also cells from downstream will be added.
.TP
\-short
The datatype of coordinate data is set to short, default is float.
.TP
.I coord_all.clm
Global coordinate file.
.TP
.I coord.clm
Coordinate file river basins should be added.
.TP
.I drainage.clm
River routing file corresponding to coord_all.clm.
.TP
.I coord_basin.clm
New coordinate file created.
.SH EXAMPLES
.B adddrain
-short
grid.clm coord.clm drainage.clm coord_basin.clm
.PP
.SH EXIT STATUS
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file


.SH SEE ALSO
lpjml(1), txt2clm(1), cru2clm(1), clm(5)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: David Hötten (davidho@pik-potsdam.de)

### Added

- Warning message added in `lpjml`, `lpjcheck`,`cdf2soil`, `txt2grid` and `cdf2coord` if cell size cannot be resolved by grid data in short format.
- Function `isfloatcoord()` added in `coord.c` to check whether coordinate must be stored as a float variable, test function `test_isfloatcoord.c` added.
- Test functions for `vmaxlimit_tree`, `vmaxlimit_grass`, `vmaxlimit_crop`, `bisect`, `leftmostzero` and `getbufferavg` functions added.

### Changed

- Error about first simulation year is less than first year of climate data replaced by warning in `initclimate.c`.
- First and last simulation year in spinup changed from 1901 to 1840 in `lpjml_config.cjson`.
- Number of spinup years changed from 3500 to 3439 in `lpjml_config.cjson` in order to get bit-identical results.
- Test files for unit test added to tar and zip file in `Makefile`.

### Removed

- Unused source file `src/lpj/part.c` removed.

### Fixed

- Expected number of bands set correctly for raw input files in `openinputdata.c`.
- Hard-coded filename `test.lpj` replaced by temporary filename in `test_append.c`, `test_getnoread.c`, `test_indexarray.c`, `test_isdefined.c`, `test_isnull.c`, `test_restart.c`, and `test_writename.c` in order to allow parallel execution of tests.
- Arguments corrected in calls of lpjml functions in `lpj_poem.c`.
- Missing `-print_noread` option added in `INSTALL`.
- `setupannual_grid.c` corrected to compile with `-DIMAGE -DCOUPLED` flag.
- Error message for different datatypes corrected in `lpcat.c`.
- Variables initialized to zero and NULL in `phenology_gsi.c`, `lpjcat.c` and `restart2yaml.c` to avoid warnings in `gcc`.
- Man page for `json2restart` corrected.
- Missing option `-nooutput` added to man page of `lpjfiles`.
- Help text for `-h` option of `printclm` corrected.
- Check for NULL pointer for `pfts` array corrected in `lpjcat.c`.
- Double entry of `getlag.$O` removed in `src/soil/Makefile`.


## [5.10.1] - 2025-09-30

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review: Maik Billing (billing@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- Variable `phen` added to output in `fprintpft.c`.
- Compile option `NRECOVERY_COST` added to `INSTALL` and man page of `configure.sh`.
- Check for valid `swc_bnf` parameters added in `fscanpftpar.c`.
- Wirth et al. (2024)  article added to `REFERENCES`.
- New parameter `"tscal_b"` added in `lpjparam.cjson`.
- New virtual function `getb()` added to type `Pft`.

### Changed

- 2 boolean flags `phen_to_one` and `rainyseason` added to tree parameter. These are used in `phenology_gsi.c` instead of string compare with tree names.
- Unused argument `aet_layer` removed from `nitrogen_stress()` function.
- `phi_bnf` parameter calculated from `swc_bnf` parameter instead of reading from `pft.cjson` file in `fscanpftpar.c`.
- Hard-coded parameter in temperature dependent calculation of `b` replaced by global parameter `tscal_b`.
- Calculation of `b` in `setupannual_grid.c` replaced by call to new virtual PFT function `getb()`.

### Removed

- Parameter `phi_bnf` removed from `pft.cjson`.
- Variable `b` removed from restart file.
- Unused macro `RESTART_VERSION` removed from `header.h`.

### Fixed

- Arguments of `npp` function corrected in `daily_woodplantation.c`. The current version leads to a too strong reduction in NPP.
- `nuptake_crop.c` changed to compile with `-DDEBUG_N` option.
- `update_daily_cell.c` changed to compile with `-DDEBUG` option.
- `daily_agriculture_tree.c` changed to compile with `-DDEBUG2` option.
- Typo in error message in `bstruct_writearrayindex.c` corrected.
- Error message corrected in `bstruct_wopen.c`.
- Error number corrected in `bstruct_finish.c`.
- Argument `temp` added to `f_turnover_tree` function to compile with `-DNRECOVERY_COST` option.


## [5.10.0] - 2025-08-14

### Contributors

- author: Jens Heinke (heinke@pik-potsdam.de), Sibyll Schaphoff (sibylls@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- option to specify year after which deforestation is no longer feeding product pools but biomass is burnt instead: `luc_burn_startyear` (disabled)
- option to specify costs for N recovery upon turnover (pre-compiler switch `NRECOVERY_COST`, disabled)

### Changed

- parameter changes:
  - `alphaa` from 0.5 to 0.7 for natural vegetation;
  - `theta` from 0.9 to 0.95;
  - `fn_turnover` PFT1: 0.5, PFT2: 0.5, PFT3: 0.7, PFT4: 0.3, PFT5: 0.3, PFT6: 0.7, PFT7: 0.3, PFT8: 0.3; updated values from [Sophia et al. 2024](https://doi.org/10.5194/bg-21-4169-2024) for all other PFTs except crops
  - `cnratio_leaf.low/median/high` for all PFTs updated from TRY data
  - `b` (ratio of dark respiration Rd to maximum photosynthetic capacity Vcmax) set to 0.031 for C3 and C4 plants (value for 25°C, from [Wang et al. 2020](https://doi.org/10.1111/gcb.14980), consistent with TRY data)

src/utils/adddrain.c
/**************************************************************************************/
/**                                                                                \n**/
/**                    a  d  d  d  r  a  i  n  .  c                                \n**/
/**                                                                                \n**/
/**     Program adds drainage basins to coordinate file                            \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

static int findroute(int index,const Routing *r,int n)
{
  int i;
  for(i=0;i<n;i++)
    if(index==r[i].index)
      return i;
  return NOT_FOUND;
} /* of 'findroute' */

int main(int argc,char **argv)
{
  Header header;
  Coordfile coordfile;
  FILE *file;
  Filename filename;
  String id,s;
  Coord *c,*c2,*cnew,res;
  float cellsize_lon,cellsize_lat;
  Routing *r;
  Bool swap,isforward,isshort;
  int index,from,to;
  int i,j,n1,n2,data_version;
  filename.fmt=CLM;
  isforward=FALSE;
  isshort=FALSE;
  for(i=1;i<argc;i++)
    if(argv[i][0]=='-')
    {
      if(!strcmp(argv[i],"-longheader"))
        filename.fmt=CLM2;
      else if(!strcmp(argv[i],"-downstream"))
        isforward=TRUE;
      else if(!strcmp(argv[i],"-short"))
        isshort=TRUE;
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n",argv[i]);
        return EXIT_FAILURE;
      } 
    }
    else
      break;
  argc-=i-1;
  argv+=i-1;
  if(argc<5)
  {
    fprintf(stderr,"Error: Missing arguments.\n"
            "Usage: %s [-longheader] [-downstream] [-short] coord_all.clm coord.clm drainage.clm coord_basin.clm\n",
            argv[1-i]);
    return EXIT_FAILURE;
  }
  filename.name=argv[1];
  coordfile=opencoord(&filename,TRUE);
  if(coordfile==NULL)
    return EXIT_FAILURE;
  n1=numcoord(coordfile);
  c=newvec(Coord,n1);
  if(c==NULL)
  {
    printallocerr("c");
    return EXIT_FAILURE;
  }
  getcellsizecoord(&cellsize_lon,&cellsize_lat,coordfile);
  res.lon=cellsize_lon;
  res.lat=cellsize_lat;
  for(i=0;i<n1;i++)
    if(readcoord(coordfile,c+i,&res))
    {
      fprintf(stderr,"Error reading cell %d in '%s'.\n",i,argv[1]);
      return EXIT_FAILURE;
    }
  closecoord(coordfile);
  filename.name=argv[2];
  coordfile=opencoord(&filename,TRUE);
  if(coordfile==NULL)
    return EXIT_FAILURE;
  n2=numcoord(coordfile);
  c2=newvec(Coord,n2);
  if(c2==NULL)
  {
    free(c);
    printallocerr("c2");
    return EXIT_FAILURE;
  }
  for(i=0;i<n2;i++)
    if(readcoord(coordfile,c2+i,&res))
    {
      fprintf(stderr,"Error reading cell %d in '%s'.\n",i,argv[1]);
      free(c);
      free(c2);
      return EXIT_FAILURE;
    }
  closecoord(coordfile);
  file=fopen(argv[3],"rb");
  if(file==NULL)
  {
    fprintf(stderr,"Error opening '%s': %s.\n",argv[3],strerror(errno));
    free(c);
    free(c2);
    return EXIT_FAILURE;
  }
  data_version=READ_VERSION;
  if(freadanyheader(file,&header,&swap,id,&data_version,TRUE))
  {
    fprintf(stderr,"Error reading header in '%s'.\n",argv[3]);
    free(c);
    free(c2);
    return EXIT_FAILURE;
  }
  if(data_version>CLM_MAX_VERSION)
  {
    fprintf(stderr,"Error: Unsupported version %d in '%s', must be less than %d.\n",
            data_version,argv[3],CLM_MAX_VERSION+1);
    free(c);
    free(c2);
    return EXIT_FAILURE;
  }
  if(header.nbands!=2)
  {
    fprintf(stderr,"Invalid number of bands=%d in '%s', must be 2.\n",header.nbands,argv[3]);
    free(c);
    free(c2);
    return EXIT_FAILURE;
  }
  if(header.nstep!=1)
  {
    fprintf(stderr,"Invalid number of steps=%d in '%s', must be 1.\n",header.nstep,argv[3]);
    return EXIT_FAILURE;
  }
  if(header.timestep!=1)
  {
    fprintf(stderr,"Invalid time step=%d in '%s', must be 1.\n",header.timestep,argv[3]);
    free(c);
    free(c2);
    return EXIT_FAILURE;
  }
  if(data_version>2 && header.datatype!=LPJ_INT)
  {
    fprintf(stderr,"Invalid datatype %s in '%s', muste bei int.\n",typenames[header.datatype],argv[3]);
    free(c);
    free(c2);
    return EXIT_FAILURE;
  }
  if(header.ncell!=n1)
  {
    fprintf(stderr,"Invalid number of cells %d in '%s', not %d.\n",
            header.ncell,argv[3],n1);
    free(c);
    free(c2);
    return EXIT_FAILURE;
  }
  r=newvec(Routing,header.ncell);
  if(r==NULL)
  {
    fclose(file);
    printallocerr("r");
    free(c);
    free(c2);
    return EXIT_FAILURE;
  }
  if(fread(r,sizeof(Routing),header.ncell,file)!=header.ncell)
  {
    fclose(file);
    free(r);
    free(c);
    free(c2);
    fprintf(stderr,"Error reading routing in '%s'.\n",argv[3]);
    return EXIT_FAILURE;
  }
  if(swap)
    for(i=0;i<header.ncell;i++)
      r[i].index=swapint(r[i].index);
  fclose(file); 
  cnew=newvec(Coord,header.ncell);
  if(cnew==NULL)
  {
    printallocerr("cnew");
    free(r);
    free(c);
    free(c2);
    return EXIT_FAILURE;
  }
  for(i=0;i<n2;i++)
    cnew[i]=c2[i];
  free(c2);
  file=fopen(argv[4],"wb");
  if(file==NULL)
  {
    fprintf(stderr,"Error creating '%s': %s.\n",argv[4],strerror(errno));
    free(cnew);
    free(r);
    free(c);
    return EXIT_FAILURE;
  }
  header.datatype=(isshort) ? LPJ_SHORT : LPJ_FLOAT;
  header.ncell=n2;
  header.cellsize_lon=cellsize_lon;
  header.cellsize_lat=cellsize_lat;
  header.firstcell=0;
  header.firstyear=0;
  header.order=0;
  header.nyear=1;
  header.nstep=1;
  header.timestep=1;
  header.nbands=2;
  header.scalar=(isshort) ? 0.01 : 1;
  for(i=0;i<header.ncell;i++)
  {
    //printf("find %g %g\n",cnew[i].lon,cnew[i].lat);
    index=findcoord(cnew+i,c,&res,n1);
    if(index==NOT_FOUND)
    {
      fprintf(stderr,"Coordinate %s not found.\n",sprintcoord(s,cnew+i));
      free(cnew);
      free(r);
      free(c);
      return EXIT_FAILURE;
    }
    for(j=0;;)
    {
      from=findroute(index,r+j,n1-j);
      if(from==NOT_FOUND)
        break;
      from+=j;
      to=findcoord(c+from,cnew,&res,header.ncell);
      if(to==NOT_FOUND) 
      {
        /* add cell from upstream */
        cnew[header.ncell]=c[from];
        header.ncell++;
      }
      j=from+1;
    }
    if(isforward)  /* add cells from downstream? */
    {
      to=r[index].index;
      if(to>=0)
      {
         index=findcoord(c+to,cnew,&res,header.ncell);
         if(index==NOT_FOUND)
         {
           cnew[header.ncell]=c[to];
           header.ncell++;
         }
      }
    }
  }
  free(r);
  free(c);
  if(fwriteheader(file,&header,LPJGRID_HEADER,LPJGRID_VERSION))
  {
    fprintf(stderr,"Error writing header in '%s'.\n",argv[4]);
    free(cnew);
    fclose(file);
    return EXIT_FAILURE;
  }
  if(isshort)
    for(i=0;i<header.ncell;i++)
      writecoord(file,cnew+i);
  else
    for(i=0;i<header.ncell;i++)
      writefloatcoord(file,cnew+i);
  fclose(file);
  if(header.ncell-n2>0)
    printf("%d cells added.\n",header.ncell-n2);
  else
    puts("No cells added.");
  free(cnew);
  return EXIT_SUCCESS;
} /* of 'main' */
